Server/src/main/java/org/gluu/oxauth/service/CibaRequestsProcessorJob.java

```diff
@@ -0,0 +1,182 @@
+
+    @Inject
+    private Event<TimerEvent> processorEvent;
+
+    @Inject
+    private CIBAPushErrorProxy cibaPushErrorProxy;
+
+    @Inject
+    private CIBAPingCallbackProxy cibaPingCallbackProxy;
+
+    @Inject
+    private AuthorizationGrantList authorizationGrantList;
+
+    @Inject
+    private CibaRequestService cibaRequestService;
+
+    private long lastFinishedTime;
+
+    private AtomicBoolean isActive;
+
+    private ExecutorService executorService;
+
+    /**
+     * Method invoked from the appInitializer to start processing every some time.
+     */
+    public void initTimer() {
+        log.debug("Initializing CIBA requests processor");
+        this.isActive = new AtomicBoolean(false);
+        int intervalSec = appConfiguration.getBackchannelRequestsProcessorJobIntervalSec();
+
+        // Schedule to start processor every N seconds
+        processorEvent.fire(new TimerEvent(new TimerSchedule(intervalSec, intervalSec),
+                new CibaRequestsProcessorEvent(), Scheduled.Literal.INSTANCE));
+
+        this.lastFinishedTime = System.currentTimeMillis();
+        this.executorService = Executors.newCachedThreadPool(ServerUtil.daemonThreadFactory());
+    }
+
+    @Asynchronous
+    public void process(@Observes @Scheduled CibaRequestsProcessorEvent cibaRequestsProcessorEvent) {
+        if (this.isActive.get()) {
+            return;
+        }
+        if (!this.isActive.compareAndSet(false, true)) {
+            return;
+        }
+
+        try {
+            if (jobIsFree()) {
+                processImpl();
+                this.lastFinishedTime = System.currentTimeMillis();
+            } else {
+                log.trace("Starting conditions aren't reached for CIBA requestes processor");
+            }
+        } finally {
+            this.isActive.set(false);
+        }
+    }
+
+    /**
+     * Defines whether the job is still in process or it is free according to the time interval defined.
+     * @return True in case it is free to start a new process.
+     */
+    private boolean jobIsFree() {
+        int interval = appConfiguration.getBackchannelRequestsProcessorJobIntervalSec();
+        if (interval < 0) {
+            log.info("CIBA Requests processor timer is disabled.");
+            log.warn("CIBA Requests processor timer Interval (cleanServiceInterval in oxauth configuration) is negative which turns OFF internal clean up by the server. Please set it to positive value if you wish internal CIBA Requests processor up timer run.");
+            return false;
+        }
+
+        long timeDiffrence = System.currentTimeMillis() - this.lastFinishedTime;
+        return timeDiffrence >= interval * 1000;
+    }
+
+    /**
+     * Main process that process CIBA requests in cache.
+     */
+    public void processImpl() {
+        try {
+            List<CIBARequest> expiredRequests = cibaRequestService.loadExpiredByStatus(
+                    CIBAGrantUserAuthorization.AUTHORIZATION_PENDING);
```

Comment:
It will work good for single oxauth. In cluster it may work as:
1) node1: req1, req2
2) node2: req3, req4
And then when job run one node will grab all requests while the other node will do nothing. Thus question, should we introduce chunkSize here ? Configurable value by which we say how many requests to grab for processing.

And second question, according to this method expired request have CIBAGrant. It somehow looks strange that we created grant for request which is not authorized. Is it by design?

Reply:
First suggestion, yeah, you are right, it's probably that one node would process the most requests. Ok, so I'll add a new json property then in order to configure it in production.

Second question, yeah, you are right, I was waiting for Javier's answer in order to process it. I'm going to create another card for that, since that's not associated with expired requests at all and this card has many changes.

Reply:
Issue created: https://github.com/GluuFederation/oxAuth/issues/1375